### ISSUE TITLE
[Feature Request] Preview toot

### ISSUE THREAD

**Issue by fallerOfFalls** [tootsuite/mastodon#1860]
On github you can preview what you've written to see how it will de displayed. This is very helpful because the formatting doesn't always come out as intended. On Mastodon: usernames, links, images, imoji art, often don't come out as intended, and there's no way to fix it other than deleting and starting again, by which time your toot may have been boosted a bunch of times. So can we have a GitHub style preview?

* * * *

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.

**Comment by thomaskuntzz** [comment]
Yeah, definitely needed! :+1: 

And I don't think the priority of this feature is low, I would rather consider it as medium.

**Comment by lipki** [comment]
I came to add this issue also.

**Comment by chaica** [comment]
I strongly disagree with the tag priority - low here. 

One or twice a week while writing a complex toot with line breaks and images, I'm asking myself what will be the result before posting and would really like the toot preview feature. It's much needed.

If entrepreneurship and startup history teach us something, that is user-oriented feature improvements (like adding a comment while boosting #309 ) or toot preview here is what will make the difference for the end user and help to gain popularity, not the backend features. Please your users with great features and they will help spread the word about Mastodon (and as you can guess I've been using the Github preview feature before posting to check if everything was fine for this message before posting).

**Comment by nightpool** [comment]
The priority-low tag is a reflection of the fact that mastodon's formatting is exceedingly simple, so any preview would match up very closely with what's already in the textbox.

images I feel like is a separate feature—we should straight up just show those on the compose form when you upload them.

**Comment by Cassolotl** [comment]
I guess it depends how you use Mastodon. The fact that emojis when selected are inserted as text for some reason, for example, annoys me a lot. So if we're not switching to emojis-in-textbox, I'd like to be able to type in what I *think* is the :emoji name: and then click preview to check I've got it right. Also folks do ASCII art as memes and I see folks mess them up, delete them and try again repeatedly until it looks right. Preview would help a lot in both of these scenarios!

**Comment by chaica** [comment]
With the toot preview you could spot if you have the right emojis, as @Cassolotl said, but the urls are also shortened and the path to mastodon user (like @carlchenet@mastodon.social) are displayed differently in the text box and in the timeline. So having the toot preview makes a real difference.

**Comment by ChameleonScales** [comment]
I also disagree on the "low priority" tag.
I have to delete and update my toots quite often and I've lost boosts several times because of it.

**Comment by charlag** [comment]
I think that it's not as useful as on GitHub because it will be different for people on different instances/clients/networks anyway?.. How it should work?..

**Comment by ChameleonScales** [comment]
for the "how should it work ?", I like the "Write" and "Preview" tabs in github. It could also be a little eye icon (as an on/off switch) right here :
![](https://i.imgur.com/WQzH6gG.png)

**Comment by charlag** [comment]
Okay, I'll try to be more clear: you want to check how this toot will look like for you? For people on your instance? For people on another instance?...

**Comment by ChameleonScales** [comment]
it doesn't make much sense to me to preview how a toot would look like on another instance than the one it's tooted on. I don't even understand how that would be possible since everthing you do on an instance is using a specific version of mastodon, so you'd have to querry the code of the version that the other instance is using. This seems very cumbersome, though I'm no expert on this.

**Comment by Cassolotl** [comment]
Ultimately people want to look at the posted toot and be like "yes this looks right to me" - you can't take responsibility for other instances, let alone other browsers and operating systems. So the preview should probably only show a preview of how it will look to the person posting it, on their instance.

**Comment by charlag** [comment]
By "another instance" I meant things like usernames in the full form etc.
Maybe I'm trying to say that it's kinda pointless: we have almost WYSIWYG with compose field, expect for emojis maybe

**Comment by chaica** [comment]
Emojis are not previewed, urls are not previewed too.

**Comment by Cassolotl** [comment]
Previews of hashtags can be good when you're accidentally putting #apostrophe's in them, or you're not sure whether hyphens work, stuff like that.

Whenever I lament about not being able to edit toots, someone *always* replies to say that instead we should do more to make sure that people don't make mistakes in the first place, so that we don't have to delete and repost so much. Toot previews would be a good step in that direction.

**Comment by charlag** [comment]
Hashtags previews is quite a problem tho. Most clients highlight tags, I think masto web should do the same but that should be another issue

**Comment by steevithak** [comment]
Yes, please! The lack of any sort of preview of a post is one my biggest complaints with Mastodon over other platforms. Essentially, I have to make a "trial" post (and sometimes the post doesn't even fully render for a few minutes after it was posted). Then if my post turns out wrong, I have to delete the post and start over. Having a real preview I could check before posting would be much better.
